langpuppet/gapi.go

```diff
@@ -0,0 +1,284 @@
+		},
+	}
+
+	dataPuppet := gapi.Data{
+		Program:       obj.data.Program,
+		Hostname:      obj.data.Hostname,
+		World:         obj.data.World,
+		Noop:          obj.data.Noop,
+		NoConfigWatch: obj.data.NoConfigWatch,
+		NoStreamWatch: obj.data.NoStreamWatch,
+		Debug:         obj.data.Debug,
+		Logf: func(format string, v ...interface{}) {
+			obj.data.Logf(puppet.Name+": "+format, v...)
+		},
+	}
+
+	if err := obj.langGAPI.Init(dataLang); err != nil {
+		return err
+	}
+	if err := obj.puppetGAPI.Init(dataPuppet); err != nil {
+		return err
+	}
+
+	obj.closeChan = make(chan struct{})
+	obj.initialized = true
+	return nil
+}
+
+// Graph returns a current Graph.
+func (obj *GAPI) Graph() (*pgraph.Graph, error) {
+	if !obj.initialized {
+		return nil, fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	var graph_err error
+
+	if obj.currentLangGraph == nil {
+		obj.currentLangGraph, graph_err = obj.langGAPI.Graph()
+		if graph_err != nil {
+			return nil, graph_err
+		}
+	}
+	if obj.currentPuppetGraph == nil {
+		obj.currentPuppetGraph, graph_err = obj.puppetGAPI.Graph()
+		if graph_err != nil {
+			return nil, graph_err
+		}
+	}
+
+	g, err := mergeGraphs(obj.currentLangGraph, obj.currentPuppetGraph)
+
+	obj.data.Logf("current graph from lang:")
+	obj.currentLangGraph.Logf(obj.data.Logf)
+	obj.data.Logf("current graph from puppet:")
+	obj.currentPuppetGraph.Logf(obj.data.Logf)
+	if err == nil {
+		obj.data.Logf("merge result:")
+		g.Logf(obj.data.Logf)
+	}
+
+	return g, err
+}
+
+// Next returns nil errors every time there could be a new graph.
+func (obj *GAPI) Next() chan gapi.Next {
+	ch := make(chan gapi.Next)
+	obj.wg.Add(1)
+	go func() {
+		defer obj.wg.Done()
+		defer close(ch) // this will run before the obj.wg.Done()
+		if !obj.initialized {
+			next := gapi.Next{
+				Err:  fmt.Errorf("%s: GAPI is not initialized", Name),
+				Exit: true, // exit, b/c programming error?
+			}
+			ch <- next
+			return
+		}
+		nextLang := obj.langGAPI.Next()
+		nextPuppet := obj.puppetGAPI.Next()
+
+		langReady := false
+		puppetReady := false
+
+		for {
+			select {
+			case nextChild := <-nextLang:
+				if nextChild.Err == nil {
+					obj.currentLangGraph = nil
+					langReady = true
+				}
+			case nextChild := <-nextPuppet:
+				if nextChild.Err == nil {
+					obj.currentPuppetGraph = nil
+					puppetReady = true
+				}
+			case <-obj.closeChan:
+				return
+			}
+
+			if !langReady || !puppetReady {
+				continue
+			}
+
+			obj.data.Logf("generating new composite graph...")
+			next := gapi.Next{
+				//Exit: true, // TODO: for permanent shutdown!
+				Err: nil,
+			}
+			select {
+			case ch <- next: // trigger a run (send a msg)
+			// unblock if we exit while waiting to send!
+			case <-obj.closeChan:
+				return
+			}
+		}
+	}()
+	return ch
+}
+
+// Close shuts down the Puppet GAPI.
+func (obj *GAPI) Close() error {
+	if !obj.initialized {
+		return fmt.Errorf("%s: GAPI is not initialized", Name)
+	}
+
+	obj.langGAPI.Close()
```

Comment:
If something returns an error, and you don't acknowledge it or pass it on, you fail. :(

Reply:
Fixed